Netable/Netable/Netable.swift

```diff
@@ -73,82 +80,93 @@ open class Netable {
+            DispatchQueue.main.async {
+                self.logDestination.log(event: event)
             }
+        }
+
+        let retryConfiguration = self.retryConfiguration
+
+        let task = urlSession.dataTask(with: urlRequest) { data, response, error in
+            let time = CACurrentMediaTime() - startTimestamp
 
             do {
                 if let error = error {
-                    self.logDestination.log(event: .requestFailed(error: NetableError.requestFailed(error)))
                     throw NetableError.requestFailed(error)
                 }
 
                 guard let response = response as? HTTPURLResponse else { fatalError("Casting response to HTTPURLResponse failed") }
+
                 guard 200...299 ~= response.statusCode else {
-                    self.logDestination.log(event: .requestCompleted(statusCode: response.statusCode, responseData: data, finalizedResult: nil))
                     throw NetableError.httpError(response.statusCode, data)
                 }
 
                 let decoded = request.decode(data)
                 switch decoded {
                 case .success(let raw):
                     let finalizedData = request.finalize(raw: raw)
-                    self.logDestination.log(event: .requestCompleted(statusCode: response.statusCode, responseData: data, finalizedResult: finalizedData))
+                    mainThreadLog(.requestCompleted(request: requestInfo, taskTime: time, statusCode: response.statusCode, responseData: data, finalizedResult: finalizedData))
                     completion(finalizedData)
                 case .failure(let error):
-                    self.logDestination.log(event: .requestFailed(error: error))
-                    completion(.failure(error))
+                    throw error
                 }
-            } catch let error as NetableError {
-                self.logDestination.log(event: .requestFailed(error: error))
-                completion(.failure(error))
             } catch {
-                let error = NetableError.decodingError(error, data)
-                self.logDestination.log(event: .requestFailed(error: error))
-                completion(.failure(error))
+                let netableError = (error as? NetableError) ?? NetableError.unknownError(error)
+
+                var isCancellation = false
+
+                if case .requestFailed(let error) = netableError {
+                    let nsError = error as NSError
+                    if nsError.domain == NSURLErrorDomain && nsError.code == NSURLErrorCancelled {
+                        isCancellation = true
+                    }
+                }
+
+                if !isCancellation && retriesLeft > 0 && retryConfiguration.enabled && retryConfiguration.errors.shouldRetry(netableError) {
+                    mainThreadLog(.requestRetrying(request: requestInfo, taskTime: time, error: netableError))
+                    DispatchQueue.main.async {
```

Comment:
Does it make more sense to move this main thread check into `delay` so it doesn't need to be copied over if we called it somewhere else?

Reply:
I went back and forth on this, the main reason I left it here is that the limitation is NOT that the DelayedOperations class REQUIRES it to be on the main thread, it just requires it to be on a thread with an active RunLoop (otherwise you can't schedule the timer) which the network thread(s) seem not to have. So although I don't think it matters in practice, it seemed more conceptually right to leave DelayedOperations able to work on any thread (with a RunLoop) and have the networking code force it onto the main thread because it doesn't have a RunLoop.

The fact that it confused you does suggest that something needs to be clearer, I think what I'll probably do it just switch it to use a DispatchSourceTimer instead, that should keep it as general purpose, without requiring any calls to be bridged to the main thread.